src/ra_server.erl

```diff
@@ -2818,6 +2889,99 @@ meta_name(#cfg{system_config = #{names := #{log_meta := Name}}}) ->
     Name;
 meta_name(#{names := #{log_meta := Name}}) ->
     Name.
+
+already_member(State) ->
+    % already a member do nothing
+    % TODO: reply? If we don't reply the caller may block until timeout
+    {not_appended, already_member, State}.
+
+%%% ====================
+%%% Voter status helpers
+%%% ====================
+
+-spec ensure_promotion_target(ra_voter_status(), ra_server_state()) ->
+    {ok, ra_voter_status()} | {error, term()}.
+ensure_promotion_target({voter, Reason}, _) ->
+    {ok, {voter, Reason}};
+ensure_promotion_target({nonvoter, #{target := _, nvid := _} = Reason}, _) ->
+    {ok, {nonvoter, Reason}};
+ensure_promotion_target({nonvoter, #{nvid := _} = Reason}, #{commit_index := CI}) ->
```

Comment:
I'm thinking using `ra_log:next_index/1` to get the promotion target would be more "correct", it should in this case be the same log index as the non voter was added at.

Reply:
Also this line could do with a newline somewhere, It goes over the 80 char mark.